src/translator/service.cpp

```diff
@@ -40,21 +63,43 @@ Service::Service(Ptr<Options> options, MemoryGift modelMemory, MemoryGift shortl
       pcqueue_(std::max<size_t>(1, numWorkers_))
 #endif
 {
+  initialize(options);
+}
 
-  if (numWorkers_ == 0) {
-    build_translators(options, /*numTranslators=*/1);
-    initialize_blocking_translator();
-  } else {
-    build_translators(options, numWorkers_);
-    initialize_async_translators();
+Service::Service(const std::string &config, const void* modelMemory, const void* shortlistMemory)
+    : numWorkers_(parseOptions(config)->get<int>("cpu-threads")),
+      text_processor_(vocabs_, parseOptions(config)), batcher_(parseOptions(config))
+#ifndef WASM
+      // 0 elements in PCQueue is illegal and can lead to failures. Adding a
+      // guard to have at least one entry allocated. In the single-threaded
+      // case, while initialized pcqueue_ remains unused.
+      ,
+      pcqueue_(std::max<size_t>(1, numWorkers_))
+#endif
+{
+  Ptr<Options> options = parseOptions(config);
+  initialize(options);
+  if(modelMemory != nullptr){
+    // MemoryGift* should be freed once it is no longer used; here is a hack to make TranslationModel works
+    const MemoryGift* modelGift = new MemoryGift(modelMemory,1);
+    modelMemory_ = modelGift;
+  }
+
+  if(shortlistMemory!= nullptr) {
+    // Hacks to obtain shortlist memory size as this will be checked during construction
+    size_t shortlistMemorySize = sizeof(uint64_t) * (6 + *((uint64_t*)shortlistMemory+4))
+            + sizeof(uint32_t) * *((uint64_t*)shortlistMemory+5);
+    // MemoryGift* should be freed once it is no longer used; here is a hack to make TranslationModel works
+    const MemoryGift* shortlistGift = new MemoryGift(shortlistMemory, shortlistMemorySize);
+    shortlistMemory_ = shortlistGift;
```

Comment:
What you're doing is `f: void* -> MemoryGift*`.

call the two `MemoryGift* hackShortList(void*)`, `MemoryGift* hackMemory(void*)` for now and move to (constructor's) member initialization list. Easy to adjust later, and we know the blocks are a hack.

This way you will have it ready before `initialize()` is called as well.

Question, not review: Do we have code to validate model-size yet?

Reply:
The following equivalency makes our life easier, does it not? 

```cpp
Service(std::string config, const void* modelMemory, 
const void *shortlistMemory)
    : Service(parseOptions(config), hackMemory(modelMemory),
    hackShortList(shortlistMemory)) {} 
```

Reply:
While we're at it, it may make sense to distill all construction routine into the most enriched one:

```cpp
Service(Ptr<Options> options):
    Service(options, nullptr, nullptr)
```

No default arguments though, because it'll end up creating many codepaths outside the API which will need change if we change stuff internally.

Reply:
> Question, not review: Do we have code to validate model-size yet?

No.

Reply:
For the comments for Service constructors, I will address in later commit (I knew current code refactoring is just a comprise but the best design).

Reply:
You are not allowed to just pass const void *.  Sizing information is required to not segfault.  We may not have code to validate model size at this time, but it should still be mandatory to pass so it can be checked later.

Reply:
Passing size is mandatory to prevent segfault.  Even if we can't validate it at this time.